ChopChop/Storage/AppDatabase.swift

```diff
@@ -329,45 +355,89 @@ extension AppDatabase {
 extension AppDatabase {
     func saveRecipe(_ recipe: inout RecipeRecord) throws {
         var ingredients: [RecipeIngredientRecord] = []
+        var graph = RecipeStepGraphRecord()
         var steps: [RecipeStepRecord] = []
-
-        try saveRecipe(&recipe, ingredients: &ingredients, steps: &steps)
+        var edges: [RecipeStepEdgeRecord] = []
+        var endPoints: [(source: RecipeStepRecord, destination: RecipeStepRecord)] = []
+
+        try saveRecipe(
+            &recipe,
+            ingredients: &ingredients,
+            graph: &graph,
+            steps: &steps,
+            edges: &edges,
+            endPoints: endPoints)
     }
 
     func saveRecipe(_ recipe: inout RecipeRecord,
                     ingredients: inout [RecipeIngredientRecord],
-                    steps: inout [RecipeStepRecord]) throws {
+                    graph: inout RecipeStepGraphRecord,
+                    steps: inout [RecipeStepRecord],
+                    edges: inout [RecipeStepEdgeRecord],
+                    endPoints: [(source: RecipeStepRecord, destination: RecipeStepRecord)]) throws {
```

Comment:
I'm not sure if the logic of turning tuples of `RecipeStepRecords` should be in the `AppDatabase` layer. I think it might be better if all logic related to transforming from our runtime models into an array of `Records` is kept in the `StorageManager` layer, and only have the `AppDatabase` layer save the records.

Reply:
The issue with this is that the edge records require the ids of the step records, which are only generated after the step records are saved. Is there another way to do this?

Reply:
I will need to think about this. At a first glance though, doesn't this implicitly mean that steps must have unique contents?. Since you're doing `let sourceStep = steps.first(where: { $0.content == source.content })`, if you have two steps with the exact same string, it won't be clear which step the edge is referring to.

Reply:
I think for this there may be no better way than to let `AppDatabase` be aware of the runtime models (e.g. have a `saveRecipe(_ recipe: Recipe)` method in `AppDatabase`. It leads to some mixing of responsibilities, but there doesn't seem to be any other way given that the method needs to be aware of the connections between each node.